### ISSUE TITLE
SameDiff: improve training and listeners

### ISSUE THREAD

**Issue by rnett** [eclipse/deeplearning4j#8049]
The SameDiff training api could use a few improvements (mostly validation data), and the current listener functionality (set for the SameDiff instance, no feedback, training only) isn't ideal.

#### Proposals:
1. `fit` methods that take in validation data and frequency, helper methods to split a `DataSetIterator` etc into training and validation data (e.g. give % validation), maybe KFold as well.  Make the results available to listeners.  Also add `fit` methods that take the validation split and use the helpers.
2. More provided listeners, use one for logging (`ScoreListener`?) and have it added by default.
3. Be able to pass additional listeners to `fit`/`eval`/`output` that are only used for that call (like how keras does it).
4. Update listener api to take in training stage (`TRAINING`/`EVALUATION`/`INFERENCE`), and to define which stages they are active for (all by default).  Also use validation results (if provided).  Listeners could also define variables they care about, which would need to be calculated each execution (e.g. if you want to report on some special value that wouldn't normally be calculated by the graph).
5. Make `epochEnd` return a boolean, if a listener returns false, stop training (for things like early stopping based on validation data).  Could do `context.stopTraining()` on a provided context, which could avoid accidents, but as that is the only thing the context would be used for using the return value is easier.
6. Add a `History` listener that is present by default and records loss values, `fit` returns this (like Keras).  Additional values could be extracted by adding additional History listeners and using the recorded values.

Maybe also add `DataSetIterators` to adjust batch size?

#### Example:
```

...
SDVariable accuracy = ...;
...

DataSetIterator data = ...;

// 75/25 training/validation split
DataSetIterator training = TrainingUtils.getTrainingData(data, 25);
DataSetIterator validation = TrainingUtils.getValidationData(data, 25);

// record accuracy every 5 epochs
History accuracyHist = new History(accuracy, 5);

// stop if accuracy gets over 95% on validation data and then decreases for 3 epochs, restoring best weights
EarlyStopping es = new EarlyStopping(accuracy, true, 0.95, 3, true);

History lossHist = SD.fit(training, 10, validation, accuracyHist, es);

graph(lossHist.recordedHistory());
graph(accuracyHist.recordedHistory());
```

**Comment by rnett** [comment]
@AlexDBlack if this looks good, I'll go ahead and implement it.

**Comment by AlexDBlack** [comment]
> fit methods that take in validation data and frequency

Sounds good

> helper methods to split a DataSetIterator etc into training and validation data (e.g. give % validation), maybe KFold as well.

No to both. Splitting and KFold is something I want in the data pipelines only, not something the net should ever have to worry about, for now at least.
We have DataSetIteratorSplitter already, but it's an ugly hack, and it can't be used for a lot of underlying iterators (anything with random order, basically... and it'll only fail at the start of the 2nd epoch when it detects it's invalid). We'll be overhaling splitting, k-fold etc in future planned DataVec work (DataVec 2.0). When we can do it properly, I'd be fine with adding a split arg.

> More provided listeners, use one for logging (ScoreListener?) and have it added by default.

Specifics? We have score and checkpoint listeners. What else?

Adding by default - no, it's opaque behaviour IMO - the "normal logging" behaviour will disappear when users set any other listeners (i.e., they won't realize it's from a listener).
And there's no way to know an appropriate frequency for different nets (1ms per iteration vs. 10 seconds per iteration - huge spam vs. never reported, for same frequency). And disabling it is opaque (have to do setListeners(null) or something)


> Be able to pass additional listeners to fit/eval/output that are only used for that call (like how keras does it).

hm... for evaluation and output, I can see that being useful.


> Update listener api to take in training stage (TRAINING/EVALUATION/INFERENCE), and to define which stages they are active for (all by default).

Sounds good. I'd put it in the "At" object rather than a new field.
https://github.com/eclipse/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/listeners/At.java

> Also use validation results (if provided).

Not following


> Listeners could also define variables they care about, which would need to be calculated each execution (e.g. if you want to report on some special value that wouldn't normally be calculated by the graph).

Hm... portability is an issue here for listeners.
There are a few valid use cases for this, so sure, sounds reasonable.


> Make epochEnd return a boolean, if a listener returns false, stop training (for things like early stopping based on validation data). Could do context.stopTraining() on a provided context, which could avoid accidents, but as that is the only thing the context would be used for using the return value is easier.

I like this. But I'm tempted to make this an enumeration:
(a) Less risk of misinterpreting what it actually does (not everyone reads the javadoc, unfortunately)
(b) We can add other behaviour (though I can't think of anything other than continue/stop... and I'd rather keep checkpoints in the dedicated listener)


> Add a History listener that is present by default and records loss values, fit returns this (like Keras). Additional values could be extracted by adding additional History listeners and using the recorded values.

Returning a history object from fit ops - sounds good. :+1:

> Maybe also add DataSetIterators to adjust batch size?

No, not yet. We'll do it properly in DataVec 2.0



As for doing evaluation in the same training loop, and reporting evaluation back to the history object: hm... fit overloads for that sound OK.

I think I'd prefer a slightly different design though, something along the lines of:

```
History hist = SD.fit(training, 10, validation, es, new Evaluation());

SomeListObject acc = hist.getEvaluation(Evaluation.Metric.ACCURACY)
Evaluation e = hist.getEvaluation(
```

SomeListObject would basically be a List<Pair<At,Double>>, but without all the generics.
Internally in the history, we'd keep the full Evaluation objects - so user can get whatever metrics they want (and stuff like ROC curves and confusion matrices - not just doubles etc)

We'll need to think about the API here a bit more though. Consider a multi-task net: you might want to do Evaluation on one output, and regression on another...

One option is do something like this:
```
History h = sameDiff.fit(training)
	.numEpochs(10)
	.testData(testSet)
	.listeners(...)
	.evaluate("softmax", labelNum, 0, new Evaluation(), new ROC())
	.fit()
```
Basically a builder for what we'd use keyword args for in python...

I also don't want an "Accuracy" SDVariable in the graph. One design difference between us and TF is that we don't add lots of random nodes to the graph that has to be stripped out later for inference/deployment - I much prefer to keep the graph structure clean all the time, and do stuff like this imperatively/externally. IMO it's easier to develop and debug, and results is less opaque behaviour when you're doing things like transfer learning or using someone else's graph (or even your own when you forgot you added training-specific nodes into it when you first trained it).
Downside of that is the additional config required.

**Comment by rnett** [comment]
>Adding by default - no, it's opaque behaviour IMO - the "normal logging" behaviour will disappear when users set any other listeners (i.e., they won't realize it's from a listener).
And there's no way to know an appropriate frequency for different nets (1ms per iteration vs. 10 seconds per iteration - huge spam vs. never reported, for same frequency). And disabling it is opaque (have to do setListeners(null) or something)

Yeah, I can see that.  For logging then, I would either use a separate interface or a subclass of of Listener, and do something like `SameDiff.setLoggers()` and `addLoggers`.

>  > Also use validation results (if provided).

> Not following

Probably better to add a listener method like `validationDone`, so that we could calculate metrics based on validation data instead of training data.

>I like this. But I'm tempted to make this an enumeration:

Agreed

For the rest of the stuff, using the graph like that was mostly an attempt to avoid having to add separate Keras Metric type stuff, but I agree with you about the graph cleanliness stuff.

One issue is that for something like EarlyStopping, we need to be able to pass in a metric to watch.  Additionally, it needs to be able to be calculated for validation or training data.  Evaluation doesn't really work (although you could make Metric a subclass of it).  Ideally this would be able to access anything in the graph, not just inputs and results like Evaluation.

My inclination would be to do something like:
```java
public interface Metric<T, R> {
    public T calculateForBatch(MultiDataSet data, SameDiff sameDiff, At at);

    public R aggregateBatches(List<T> batchResults);
}

public abstract class DoubleMetric implements Metric<Double, Double> {
    public AggregationMethod aggregationMethod(){
        return AggregationMethod.Average;
    }

    public Double aggregateBatches(List<Double> batchResults){
        return aggregationMethod().aggregate(batchResults);
    }
}
```

Metrics could be added by listeners (`Listener.requiredMetrics()`) or directly added passed to the SameDiff methods (I would probably have a common interface between them and IEvaluation so we could use one vararg param for both).

Listener methods could then receive a `Map<String, Object>` of metric results, and `History` would record all metrics and evaluations.


Also, unless I misread things, Evaluation doesn't have a defined score type method, so doing something like `history.getEvaluation(Accuracy)` would be impossible without getting the specific evaluation first.  

Imo it might be better to keep Evaluation as a separate "evaluate after training" thing rather than using it in `fit`.  The only use I can think of would be to see something like a confusion matrix after every epoch.

**Comment by AlexDBlack** [comment]
Implemented long ago
